### ISSUE TITLE
detect fan mode and heat/cool mode

### ISSUE THREAD

**Issue by stleusc** [iprak/sensi#10]
I have 4 Sensi in my house. 3 are for heating, 1 for cooling.

The cooling supports:
```
_attr_hvac_modes = [
    HVACMode.COOL,
    HVACMode.OFF,
]
_attr_supported_features = (
    ClimateEntityFeature.TARGET_TEMPERATURE | ClimateEntityFeature.FAN_MODE
)      
_attr_fan_modes = [SENSI_FAN_AUTO, SENSI_FAN_ON]  
```

Heating however only supports (it's baseboard heating with hot water boiler):
```
_attr_hvac_modes = [
    HVACMode.HEAT,
    HVACMode.OFF
]
_attr_supported_features = (
    ClimateEntityFeature.TARGET_TEMPERATURE
)
```

I manually patched your code to support that.
I remember there is a way to detect if heating and/or cooling is supported, I just don't remember how.

So my current patch is to switch between those 2 sets by checking the name of my thermostat.
Obviously that is not a generic solution. And I have to patch your code every time.

But it is rather annoying to see modes that are not supported.

Happy to help implement if needed.
Gettting this added would be AMAZING!

**Comment by iprak** [comment]
Can you let me know what are your thermostat setting? I referenced [this page](https://sensi.copeland.com/en-us/support/how-do-i-configure-my-thermostat) but I wasn't sure what heating mode you used.

**Comment by stleusc** [comment]
in the state response I can get the following:
capabilities.cool_cycle_rate: no 
capabilities.heat_cycle_rate: yes
for my heating.
capabilities.cool_cycle_rate: yes 
capabilities.heat_cycle_rate: no
for my cooling.

That should easily allow to detect if heat and/or cool is supported.

Even better, I am getting

capabilities.operating_mode_settings for my heating:
```json
"operating_mode_settings": {
  "off": "yes",
  "heat": "yes",
  "cool": "no",
  "aux": "no",
  "auto": "no"
}
```

and capabilities.operating_mode_settings for my cooling:
```json
"operating_mode_settings": {
  "off": "yes",
  "heat": "no",
  "cool": "yes",
  "aux": "no",
  "auto": "no"
},
```

When connecting to the websocket I am using:
```
capabilities=display_humidity,operating_mode_settings,fan_mode_settings,indoor_equipment,outdoor_equipment,indoor_stages,outdoor_stages,continuous_backlight,degrees_fc,display_time,keypad_lockout,temp_offset,compressor_lockout,boost,heat_cycle_rate,heat_cycle_rate_steps,cool_cycle_rate,cool_cycle_rate_steps,aux_cycle_rate,aux_cycle_rate_steps,early_start,min_heat_setpoint,max_heat_setpoint,min_cool_setpoint,max_cool_setpoint,circulating_fan,humidity_control,humidity_offset,humidity_offset_lower_bound,humidity_offset_upper_bound,temp_offset_lower_bound,temp_offset_upper_bound
```

Obviously not all is needed. I think I threw the kitchen sink at it to get the most information I could get.

Still looing into the circulating fan/fan setting since the API return that I support that on ALL thermostats.
However the heating is a boiler, clearly no fan.
```json
"fan_mode_settings": {
  "auto": "yes",
  "on": "yes",
  "smart": "no"
},
```
This could/should be used by you, but in my case it is wrong for my heating. I will follow up on that. Maybe I setup my thermostat wrong.

**Comment by stleusc** [comment]
Current conclusion is that Sensi does NOT properly support setting up a system without a fan.
I am wondering if there could be a way to add something into the add-on to make it a configuration like:
```
heating:
  support_fan: false
```
That way I could remove the fan option. It is annoying that this is basically a Sensi bug that we need to work around.

I know, lot's of info!

**Comment by iprak** [comment]
I pushed out a new release which should resolve this.

**Comment by stleusc** [comment]
I like what I am seeing....
However, now all 4 are actually heating only. For whatever reason, you detect my AC one as heating also.
LOVE the fan toggle!

Regarding entity naming, I see you changed that as well.

One idea for that, could you use the name that I gave the thermostat instead?

The API returns it under:
```
["state",[{"icd_id":"xxxx","registration":{"city":"yyyy","name":"Basement","
```

If you could use that name instead of the icd_id that would make using the entity nicer. Of course, clean the name (replace space with underscore, convert to all lower case and if needed remove all illegal characters (keep letters, numbers, nothing else).

So one suggestion, and one bug report.

Unsure why my COOL only is detected as HEAT only.

**Comment by stleusc** [comment]
From the debug log:
```
2023-09-14 08:10:12.815 DEBUG (MainThread) [custom_components.sensi] Basement Capabilities={<Capabilities.CONTINUOUS_BACKLIGHT: 'continuous_backlight'>: True, <Capabilities.DEGREES_FC: 'degrees_fc'>: True, <Capabilities.DISPLAY_HUMIDITY: 'display_humidity'>: True, <Capabilities.DISPLAY_TIME: 'display_time'>: True, <Capabilities.CIRCULATING_FAN: 'circulating_fan'>: True, <Capabilities.FAN_MODE_AUTO: 'fan_mode_settings.auto'>: True, <Capabilities.FAN_MODE_ON: 'fan_mode_settings.on'>: True, <Capabilities.OPERATING_MODE_OFF: 'operating_mode_settings.off'>: True, <Capabilities.OPERATING_MODE_HEAT: 'operating_mode_settings.heat'>: True, <Capabilities.OPERATING_MODE_COOL: 'operating_mode_settings.cool'>: False, <Capabilities.OPERATING_MODE_AUTO: 'operating_mode_settings.auto'>: False}
2023-09-14 08:10:12.816 DEBUG (MainThread) [custom_components.sensi] Cooling Capabilities={<Capabilities.CONTINUOUS_BACKLIGHT: 'continuous_backlight'>: True, <Capabilities.DEGREES_FC: 'degrees_fc'>: True, <Capabilities.DISPLAY_HUMIDITY: 'display_humidity'>: True, <Capabilities.DISPLAY_TIME: 'display_time'>: True, <Capabilities.CIRCULATING_FAN: 'circulating_fan'>: True, <Capabilities.FAN_MODE_AUTO: 'fan_mode_settings.auto'>: True, <Capabilities.FAN_MODE_ON: 'fan_mode_settings.on'>: True, <Capabilities.OPERATING_MODE_OFF: 'operating_mode_settings.off'>: True, <Capabilities.OPERATING_MODE_HEAT: 'operating_mode_settings.heat'>: False, <Capabilities.OPERATING_MODE_COOL: 'operating_mode_settings.cool'>: True, <Capabilities.OPERATING_MODE_AUTO: 'operating_mode_settings.auto'>: False}
```

We know the fan modes are messed up and your amazing toggle fixes that for me.
But you can see that one clearly shows Capabilities.OPERATING_MODE_HEAT as true and the other as false.

However, for whatever reason they both only support heat mode in HA.
At first glance I did not see the bug in your code but I also only looked at the changes, not the overall code.

I will see if I quickly spot what is causing this.

Regarding entity id. I am basically suggesting to replace the icd_id with the name like you had it before. Adding the name to the sub entities like humidity should stay! That was a huge important change.
At the end, that just makes it easier in yaml automations or node red because the entitiy name tells me what it is.
It does not impact functionality at all. So ultimately that is your choice.

Thanks for the hard work! I know how it is to work on a project like that....
I appreciate all you do for it!

**Comment by stleusc** [comment]
added the following debug lines to climate.py

```
    @property
    def hvac_modes(self) -> list[HVACMode]:
        """Return the list of available hvac operation modes."""

        modes = []

        if self._device.supports(Capabilities.OPERATING_MODE_OFF):
            modes.append(HVACMode.OFF)
        if self._device.supports(Capabilities.OPERATING_MODE_HEAT):
            modes.append(HVACMode.HEAT)
        if self._device.supports(Capabilities.OPERATING_MODE_COOL):
            modes.append(HVACMode.COOL)
        if self._device.supports(Capabilities.OPERATING_MODE_AUTO):
            modes.append(HVACMode.AUTO)

        LOGGER.info("%s MODES update: %s", self._device.name, modes)
        LOGGER.info("%s id %d, MODES capabilities id %d=%s", self._device.name, id(self._device), id(self._device._capabilities), self._device._capabilities)
        return modes
```

The id() function returns an id that is UNIQUE per object, so clearly we would expect id(self._device) AND id(self._device._capabilities) to be unique per device.
My debug log however shows a unique id for the device but the SAME id for the capabilities on all 4 devices.
Somewhere in the code the dictionary _capabilities is the same object assigned to all devices. Meaning, I could change one value and it will show up in all 4 devices.

I will dig more into it. Unless you see the issue it might be easier for me to find it since I have the multiple devices on my account.

I'll keep you posted!

**Comment by iprak** [comment]
Hmm I too will dig more. I have been testing with 2 devices and only one has cooling off.

As for your other question, all the DataCoordinator entities update on schedule and that is when the new fan_support setting will get picked. I think it should be possible to force the coordinator to refresh on fan_support change. But I think this will not be a common workflow.

**Comment by iprak** [comment]
Using name in entity_id should also be possible but I wonder what will happen in HA if one changes the name from the app.

**Comment by stleusc** [comment]
I found the issue... 

```
class SensiDevice:
    """Class representing a Sensi thermostat device."""

    # pylint: disable=too-many-instance-attributes
    # These attributes are meant to be here.

    coordinator = None

    identifier: str | None = None
    name: str | None = None
    model: str | None = None

    temperature: float | None = None
    temperature_unit = UnitOfTemperature.FAHRENHEIT
    humidity: int | None = None
    hvac_mode: HVACMode = HVACMode.AUTO

    _display_scale = "f"
    """Raw display_scale"""

    _capabilities: dict[Capabilities, bool] = {}
    _properties: dict[Settings, StateType] = {}
...
```

I am a Java/C developer and basically never used Python. But from a Google search it seems like every property defined OUTSIDE of __init__ is considered a class variable, owned by the class (static in Java) and NOT the instance itself.

That explains why the id() is the same for every instance. It seems like pylint agrees with my statement based on the comment in your code.

Now, I think that also means that ALL the other properties would have the same issue. One reason why it might not show could be that your code updates the value, then uses it. Then it moves to the next devices, updates, uses, etc. In that case, even though the value would change for ALL devices, it might be the case that you never use it unless you update it with new data.

In any case, chaning __init__ like this:

```
    def __init__(self, coordinator, data_json: dict) -> None:
        """Initialize a Sensi thermostate device."""
        self.coordinator = coordinator
        self._capabilities = {}
        self.update(data_json)
```
Fixes the issue and now my debug code prints a unique ID per device and my cooling device works.
HOWEVER, I believe that probably ALL your properties should be initialized in __init__ based on my reading regarding class variable (what you have) vs instance variable (what you want).

I have a feeling that many of the attributes you currently define outside of __init__ should be moved inside of __init__ and that it is simply the fact that you run update then use per device what masks the issue.

**Comment by stleusc** [comment]
> Hmm I too will dig more. I have been testing with 2 devices and only one has cooling off.
> 
> As for your other question, all the DataCoordinator entities update on schedule and that is when the new fan_support setting will get picked. I think it should be possible to force the coordinator to refresh on fan_support change. But I think this will not be a common workflow.

I thought that it simply was not pushed out. Since I found the bug (see above) you can totally ignore this. There is no need to push it out. That part is working. Disregard that thought!

**Comment by stleusc** [comment]
> Using name in entity_id should also be possible but I wonder what will happen in HA if one changes the name from the app.

I think that still should be fine. Did you not use the name up until recently? I did not updated for a few versions since I needed my own patch. So I don't know when you made the change. But in V1.0.0 I am sure the entity was named after the device name.

But if you think it will cause issues, then keep it. It's more a nice to have if it works. But if it messes things up then keep what you have.

What I am wondering though is, what if I change the device ID in HA. Does that impact your code?

**Comment by iprak** [comment]
You are absolutely right. I remembered this but I too am more of a C# developer and lost track. I will try to get this straightened out tonight or soon. Thanks for debugging into it.

**Comment by stleusc** [comment]
coordinator, update() function...

tweaked the end LOGGER part like this:

```
            LOGGER.info("%s %d %d %d", self.name, id(self.temperature), id(self._properties), id(self.attributes)),
            # pylint: disable=line-too-long
            LOGGER.info(
                "%s: %d%s humidity=%d hvac_mode=%s fan_mode=%s hvac_action=%s cool_target=%d heat_target=%d",
                self.name,
                self.temperature,
                self.temperature_unit,
                self.humidity,
                self.hvac_mode,
                self.fan_mode,
                hvac_action,
                self.cool_target,
                self.heat_target,
            )
```

And it get the following output:

```
2023-09-14 09:59:58.394 INFO (MainThread) [custom_components.sensi] Cooling 139901011117296 139901298595904 139901286000384
2023-09-14 09:59:58.414 INFO (MainThread) [custom_components.sensi] Basement 139902021331144 139901298595904 139901286000384
```

You can see that self.temperature is a unique object per device instance (not sure why, since defined outside of __init__). But you can also see that self._properties AND self.attributes are shared between ALL devices.

Almost seems like that all the 'simple types' like float, str, etc magically are per device even though my quick reading tells me it should not be.

However ALL the dict definitly are class variables unless you move them to __init__.

**Comment by stleusc** [comment]
> You are absolutely right. I remembered this but I too am more of a C# developer and lost track. I will try to get this straightened out tonight or soon. Thanks for debugging into it.

My pleasure... after all, initiated this idea hahaha.

In summary, all dict should be within __init__!
In my mind, the rest should be too (unless you want it to be shared between devices).

**Comment by stleusc** [comment]
I tried to find a change in your code where you switched the entity id from the name to the id.
Maybe I changed the ID manually in HA.
To not break things for everyone else, I'd say, keep it as you have it now.
It did break all my IDs but I can fix that. Better one person than everyone.

Thanks for the hard work! Finally I can use my Sensi in the way I like and it works perfectly!

**Comment by iprak** [comment]
> However ALL the dict definitly are class variables unless you move them to **init**.

Yup that is the gist. Initially I had `simple` class variables defaulted to some value, they would get set in update and were all fine. I followed the same declaration for the dict which became a problem. Unfortunately, I also only had one thermostat till you brought up your case so never saw the issue.

**Comment by iprak** [comment]
> I tried to find a change in your code where you switched the entity id from the name to the id. Maybe I changed the ID manually in HA. To not break things for everyone else, I'd say, keep it as you have it now. It did break all my IDs but I can fix that. Better one person than everyone.
> 
> Thanks for the hard work! Finally I can use my Sensi in the way I like and it works perfectly!

Ok, I will keep the entity_id as it is. I was following the pattern in HomeAssistant where entities would have deviceId as part of it so keep it unique. The entity_id can still be changed from the gear icon.

**Comment by iprak** [comment]
> > You are absolutely right. I remembered this but I too am more of a C# developer and lost track. I will try to get this straightened out tonight or soon. Thanks for debugging into it.
> 
> My pleasure... after all, initiated this idea hahaha.
> 
> In summary, all dict should be within **init**! In my mind, the rest should be too (unless you want it to be shared between devices).

I think it is fine to define class variable as None. Most of the other are None or have a default value. They all get set to something and at that point they become instance variables.

**Comment by iprak** [comment]
> > Using name in entity_id should also be possible but I wonder what will happen in HA if one changes the name from the app.
> 
> I think that still should be fine. Did you not use the name up until recently? I did not updated for a few versions since I needed my own patch. So I don't know when you made the change. But in V1.0.0 I am sure the entity was named after the device name.
> 
> But if you think it will cause issues, then keep it. It's more a nice to have if it works. But if it messes things up then keep what you have.
> 
> What I am wondering though is, what if I change the device ID in HA. Does that impact your code?

Device_id comes from Sensi and would not change. May you are referring to entity_id and changing that should be fine. HomeAssistant takes care of it, yes you would have to update your cards to use the new entity_id.

**Comment by iprak** [comment]
@stleusc  Do you want to give my merge (https://github.com/iprak/sensi/pull/16) a try?

**Comment by stleusc** [comment]
Installed it and gave it a quick glance... All I see is PERFECTION :-)
I'll check it out more tonight but the modes are correct and I love the entity IDs.

More feedback later.

**Comment by stleusc** [comment]
Only requires one word as feedback:

PERFECT!

Thank you so much for this change. 
If I'm ever in your neck of woods I'd be happy to buy you a beer (or two).
I think you can close this once the change is merged.

**Comment by stleusc** [comment]
Closing this issue. Works perfectly with the latest release.
Thanks again!
